### PR TITLE
Struct parser improvements / bug fixes

### LINKED ISSUE #4: Float Values
When I parse a struct with a float value, it truncates the structure, returning only the value before the decimal. For example

parseStruct([{major=Accounting, gpa=2.99}]) will return 

{
    major: Accounting,
    gpa: 2
}

### PR DESCRIPTION
**Pull request by ghost**
This PR resolves all current issues:
- #1 
- #4 
- #5 

Changes:
- The major issue #5 when nested structs in place was fixed by replacing regex:
`const toKeyValRe = /((?:\[)[^\]]*(?:\])|(?:(?:(?:},\s*?[^=]{)|,|^)\s*?\w*=))/` with function `toKeyVal`
The regex is broken for nested structs and at this point it's really hard to follow and debug the regex itself. I spend few hours tho was not able to "fix" the regex to avoid nested struct format corruption. In my opinion replacing it with the function brings more clarity and obviously fixes the issue.
- deprecated `substr` function was replaced with `substring` function
- `value * 1` an elegant way to convert a string to a number (either int or float), resolves #4

### CODE REVIEW

**Comment by windhamg** [comment]
This PR works great, except that arrays containing more than 1 item are not parsed correctly. The patch below fixes this.

```
diff --git a/src/index.js b/src/index.js
index 9e8ebd4..cc7fd04 100644
--- a/src/index.js
+++ b/src/index.js
@@ -3,6 +3,7 @@ const looseStructRe = /{.*?(?:(?:(?:},\s*?{)|,|{|)\s*?(?:\w+)=).*?}/
 
 const toKeyVal = (struct) => {
   let insideNestedStruct = 0
+  let insideArray = 0
   let structToKeyVal = []
   let startSlice = 0
   let keyValueSingePair = []
@@ -15,8 +16,12 @@ const toKeyVal = (struct) => {
       insideNestedStruct++
     } else if (struct[i] === "}") {
       insideNestedStruct--
+    } else if (struct[i] === "[") {
+      insideArray++
+    } else if (struct[i] === "]") {
+      insideArray--
     }
-    if (insideNestedStruct === 0) {
+    if (insideNestedStruct === 0 && insideArray == 0) {
       if (struct[i] === ",") {
         // the way how we get "key=value"
         keyValueSingePair = struct.slice(startSlice, i)
```

**Comment by ghost** [comment]
@windhamg could you please share an example when this is not working?

**Comment by windhamg** [comment]
Sure, if you do something like this (before my patch is applied):
```
console.log(parseStruct('{foo=bar, baz=[moe, larry, curly]}'))
```
You'll get:
```
{
  baz: "[moe",
  foo: "bar",
  larry: "curly]"
}
```

**Comment by ghost** [comment]
@windhamg thanks, the patch is merged to the pr

**Comment by rooty0** [comment]
my account got deleted, the PR https://github.com/jutaz/athena-struct-parser/pull/7
